File: app/src/main/java/com/trevorwiebe/trackacow/domain/dataLoaders/main/cow/DeleteCow.java
package com.trevorwiebe.trackacow.domain.dataLoaders.main.cow;

import android.content.Context;
import android.os.AsyncTask;

import com.trevorwiebe.trackacow.data.db.AppDatabase;
import com.trevorwiebe.trackacow.data.db.entities.CowEntity;

public class DeleteCow extends AsyncTask<Context, Void, Void> {

    private CowEntity cowEntity;

    public DeleteCow(CowEntity cowEntity){
        this.cowEntity = cowEntity;
    }

    @Override
    protected Void doInBackground(Context... contexts) {
        AppDatabase.getAppDatabase(contexts[0]).cowDao().deleteCow(cowEntity);
        return null;
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/domain/use_cases/cow_use_cases/DeleteCow.kt
package com.trevorwiebe.trackacow.domain.use_cases.cow_use_cases

import android.app.Application
import com.trevorwiebe.trackacow.data.mapper.toCacheCowModel
import com.trevorwiebe.trackacow.domain.models.cow.CowModel
import com.trevorwiebe.trackacow.domain.repository.local.CowRepository
import com.trevorwiebe.trackacow.domain.repository.remote.CowRepositoryRemote
import com.trevorwiebe.trackacow.domain.utils.Constants
import com.trevorwiebe.trackacow.domain.utils.Utility

class DeleteCow(
    private val cowRepository: CowRepository,
    private val cowRepositoryRemote: CowRepositoryRemote,
    private val context: Application
) {

    suspend operator fun invoke(cowModel: CowModel) {

        cowRepository.deleteCow(cowModel)

        if (Utility.haveNetworkConnection(context)) {
            cowRepositoryRemote.deleteCowRemote(cowModel)
        } else {
            Utility.setNewDataToUpload(context, true)
            cowRepository.insertCacheCow(cowModel.toCacheCowModel(Constants.DELETE))
        }

    }
}